### PR TITLE
Adds support for Nextcloud Deck, a Kanban-Style task management tool …

### PR DESCRIPTION
**Pull request by lenalebt**
…integrated with Nextcloud

It already works more or less, but I'm not very used to writing python code. I'm open for suggestions on how to improve stuff :-). I think functionality-wise it should all be there already.

A bit of context: I'm using nextcloud deck for managing stuff in the household together with my wife, but I'd like to get those tasks into my taskwarrior as well ;-). I currently plan to do this regularly in the upcoming time.

### CODE REVIEW

File: bugwarrior/docs/services/deck.rst
Comment (ryneeverett) [inline bugwarrior/docs/services/deck.rst]: I would move this "Attention" sentence into [a note](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html) using `..note::`.

File: bugwarrior/services/deck.py
Comment (ryneeverett) [inline bugwarrior/services/deck.py]: Regardless of the type of password, isn't it a required field? If so a default value should not be specified.

File: bugwarrior/services/deck.py
Comment (ryneeverett) [inline bugwarrior/services/deck.py]: I suspect we should use `config.StrippedTrailingSlashUrl` here so if somebody adds an extra slash at the end in their config it doesn't break url construction.

File: bugwarrior/services/deck.py
Comment (ryneeverett) [inline bugwarrior/services/deck.py]: I think it would be more clear to express this as `self.record['assignedUsers'][0]['participant']['uid'] if self.record['assignedUsers'] else None`

File: bugwarrior/services/deck.py
Comment (ryneeverett) [inline bugwarrior/services/deck.py]: I would like to see this complex logic replaced with a filter function. Many of the services have a method called `filter_repos` you could see for reference. (You'd probably call yours `filter_boards`.)

**Comment by lenalebt** [comment]
Thanks for the review, highly appreciated and very good points raised :smile: !

I would love to test the `filter_repos`, but had some trouble setting the config values needed for that. Is there an easy way to do so? Example code:

    def test_filter_repos_include(self):
        self.service.config['deck.include_board_ids'] = ['5']
        self.assertTrue(self.service.filter_repos({'title': 'testboard', 'id': 5}))
        self.assertFalse(self.service.filter_repos({'title': 'testboard', 'id': 6}))

    def test_filter_repos_exclude(self):
        self.service.config['deck.exclude_board_ids'] = ['5']
        self.assertFalse(self.service.filter_repos({'title': 'testboard', 'id': 5}))
        self.assertTrue(self.service.filter_repos({'title': 'testboard', 'id': 6}))

I think I'm just holding it wrong. It's not that I think those are absolutely crucial tests, but it would be nice to have it working somehow.

**Review by ryneeverett — APPROVED** [review]

**Comment by ryneeverett** [comment]
> I would love to test the `filter_repos`, but had some trouble setting the config values needed for that. Is there an easy way to do so?

We don't seem to have the test infrastructure to do this as trivially as would be ideal, but you can check out `test_gitlab.py:TestGitlabService` for an example. If you copied the `setUp` method and altered it to have the minimum configuration for deck, you could then test the configurations with:

```py
def test_filter_repos_include(self):
    self.config.set('myservice', 'deck.include_board_ids', '5')
    self.assertTrue(self.service.filter_repos({'title': 'testboard', 'id': 5}))
    self.assertFalse(self.service.filter_repos({'title': 'testboard', 'id': 6}))

def test_filter_repos_exclude(self):
    self.config.set('myservice', 'deck.exclude_board_ids', '5')
    self.assertFalse(self.service.filter_repos({'title': 'testboard', 'id': 5}))
    self.assertTrue(self.service.filter_repos({'title': 'testboard', 'id': 6}))
```

**Comment by lenalebt** [comment]
I tried doing that, but I had some trouble initially with the setup because all the mocking works differently over there. I worked it out - thanks for your help :-).

Besides that I fixed the validation problems around line lengths. I'm happy with it now, if you're fine with it it can be merged.

File: bugwarrior/services/deck.py
Comment (ryneeverett) [inline bugwarrior/services/deck.py]: I believe you should be able to remove the `only_if_assigned` field without consequence. It is defined in `ServiceConfig` from which this class inherits.

File: bugwarrior/services/deck.py
Comment (ryneeverett) [inline bugwarrior/services/deck.py]: Were you planning to complete annotations support? I believe every other service currently has this feature.

**Review by ryneeverett — APPROVED** [review]

File: bugwarrior/services/deck.py
Comment (lenalebt) [reply bugwarrior/services/deck.py]: Correct, thanks!

File: bugwarrior/services/deck.py
Comment (lenalebt) [reply bugwarrior/services/deck.py]: I did not initially, but I just added it. It does slow down the process significantly, though, because there is a specific call needed per card / issue. For my cases it's about factor 10 (20 seconds instead of 2). As far as I can see it, it's not possible to skip annotation importing via configuration somehow, correct? How would you handle that - it annoys me quite a lot if it slows down the process considerably, but I also see value in completeness.

Comment (ryneeverett) [reply bugwarrior/services/deck.py]: This is typically handled by setting the main configuration option "annotation_comments" to `False`. When this is set `IssueService.build_annotations` just throws the annotations away, so several services have implemented logic to avoid calling the `client.get_comments` method when when this is the case. (You can just pass an empty list to `build_annotations`. You can see the github and gitlab services for examples.)

I don't believe we have any way to disable annotations on an individual service basis or that anyone has requested that.

**Review by ryneeverett — APPROVED** [review]

File: bugwarrior/services/deck.py
Comment (ryneeverett) [inline bugwarrior/services/deck.py]: Looks like you left out `self.api_base_path`. I would combine them in the `__init__` method. (`self.api_base_path = f'{base_uri/index.php/apps/deck/api/v1.0'`) Note that there's also no need to assign any of the `__init__` arguments to `self` if you take this approach as they're only used in the `__init__` method.

**Review by ryneeverett — CHANGES_REQUESTED** [review]

File: bugwarrior/services/deck.py
Comment (ryneeverett) [reply bugwarrior/services/deck.py]: Oh nevermind, i missed that it is a totally different api path.

**Review by ryneeverett — APPROVED** [review]

**Comment by ryneeverett** [comment]
@lenalebt I hope you don't mind I pushed a couple commits to your branch to fix CI and implement conditional annotation fetching. Setting `annotation_comments = False` in the `[general]` section of your bugwarriorrc should satisfy your needs.

**Comment by lenalebt** [comment]
Oh wow, no problem, thank you! I had a little bit too much on my table lately.

So, I'm fine with merging now, I'm using it on a regular basis and don't see problems so far. Anything I can do now in addition to get it merged :-)?

**Comment by ryneeverett** [comment]
> So, I'm fine with merging now, I'm using it on a regular basis and don't see problems so far. Anything I can do now in addition to get it merged :-)?

Nope, I was just waiting for your approval of the changes. :)

**Pull request merged**
